### PR TITLE
crimson: drop OmapIterator, use omap_get_values() instead

### PR DESCRIPTION
**Pull request by rzarzynski**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/replicated_recovery_backend.cc [hunk 1]
```diff
@@ -538,43 +554,38 @@ ReplicatedRecoveryBackend::read_omap_for_push_op(
...
-        if (const uint64_t entry_size =
-            omap_iter->key().size() + omap_iter->value().length();
+        if (const uint64_t entry_size = key.size() + value.length();
             entry_size > *max_len) {
-          *max_len -= entry_size;
+          stop = true;
+          break;
         } else {
-          *max_len = 0;
+          *max_len -= std::min(*max_len, entry_size);
```
Comment (rzarzynski) [inline src/crimson/osd/replicated_recovery_backend.cc:574]: Might be worth `assert(max_len)` or switching from pointer no reference.

File: src/crimson/osd/replicated_recovery_backend.cc [hunk 2]
```diff
@@ -538,43 +554,38 @@ ReplicatedRecoveryBackend::read_omap_for_push_op(
...
         }
-        return omap_iter->next();
-      });
-    });
+        push_op->omap_entries.emplace(key, value);
+      }
+      if (!push_op->omap_entries.empty()) {
+        // we iterate in order
+        new_progress.omap_recovered_to = std::rbegin(push_op->omap_entries)->first;
```
Comment (rzarzynski) [inline src/crimson/osd/replicated_recovery_backend.cc:580]: This is also used as the input parameter to `omap_get_values()`.

File: src/osd/PGLog.cc [hunk 3]
```diff
@@ -1146,12 +1146,23 @@ namespace {
...
 
-      return store.get_omap_iterator(ch, pgmeta_oid).then([this](auto iter) {
-        return seastar::do_until([iter] { return !iter->valid(); },
-                                 [iter, this]() mutable {
-          process_entry(iter);
-          return iter->next();
-        });
+      return seastar::repeat([ch=std::move(ch),
+                              pgmeta_oid=std::move(pgmeta_oid),
+                              start=std::make_optional<std::string>(),
```
Comment (rzarzynski) [inline src/osd/PGLog.cc:1151]: Modifying this is the reason for `mutable`.

File: src/osd/PGLog.cc [hunk 4]
```diff
@@ -1146,12 +1146,23 @@ namespace {
...
+                              this]() mutable {
+        return store.omap_get_values(
+          ch, pgmeta_oid, start
+        ).safe_then([this, &start](auto ret) mutable {
```
Comment (rzarzynski) [inline src/osd/PGLog.cc:1155]: No need to copy the map.

File: src/crimson/osd/replicated_recovery_backend.cc (hunk 1 above)
Comment (rzarzynski) [reply src/crimson/osd/replicated_recovery_backend.cc]: Switched to ref.

**Review by athanatos — APPROVED** [review]

**Comment by rzarzynski** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/replicated_recovery_backend.cc:560:19: error: lambda capture 'progress' is not used [-Werror,-Wunused-lambda-capture]
    ).safe_then([&progress, &new_progress, &max_len, push_op](const auto& ret) {
```

Will fix.

**Comment by rzarzynski** [comment]
Fixed. No other changes, just killing the unused capture.

**Comment by liu-chunmei** [comment]
jenkins test make check arm64

File: src/crimson/osd/replicated_recovery_backend.cc [hunk 5]
```diff
@@ -538,43 +554,38 @@ ReplicatedRecoveryBackend::read_omap_for_push_op(
...
-        }
-        if (push_op->omap_entries.empty()) {
-          return false;
+  return seastar::repeat([&progress, &new_progress, max_len, push_op, &oid, this] {
+    return shard_services.get_store().omap_get_values(
+      coll, ghobject_t{oid}, nullopt_if_empty(new_progress.omap_recovered_to)
+    ).safe_then([&progress, &new_progress, max_len, push_op](auto ret) {
+      const auto& [done, kvs] = ret;
+      bool stop = done;
+      // assuming "values.empty() only if done" holds here!
```
[the remark this answers is not in the document]
Comment (liu-chunmei) [reply src/crimson/osd/replicated_recovery_backend.cc:563]: what's the problem of omap iterator? why change it?

[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/osd/replicated_recovery_backend.cc:563]: The synchronous version in classic is pretty convenient, but the async version in crimson is both inefficient and complex to use.

**Pull request merged**